LISA/LISAImplementation.cpp

```diff
@@ -577,20 +578,31 @@ class LISAImplementation : public Exchange::ILISA {
 
         std::vector<LISA::DataStorage::AppDetails> appsDetailsList{};
         auto rc = executor.GetAppDetailsList(type, id, version, appName, category, appsDetailsList);
-        std::map<std::pair<std::string, std::string>, std::list<AppVersionImpl*>> appsDet;
-        for(const auto& app: appsDetailsList)
-        {
-            appsDet[{app.type, app.id}].push_back(Core::Service<AppVersionImpl>::Create<AppVersionImpl>(app.version, app.appName, app.category, app.url));
-        }
-        std::list<AppImpl*> apps;
-        for(const auto& app: appsDet)
-        {
-            apps.push_back(Core::Service<AppImpl>::Create<AppImpl>(app.first.first, app.first.second, app.second));
+        if (rc == Core::ERROR_NONE) {
+
+            std::map<std::pair<std::string, std::string>, std::list<AppVersionImpl*>> appsDet;
+            for(const auto& app: appsDetailsList)
+            {
+                appsDet[{app.type, app.id}].push_back(Core::Service<AppVersionImpl>::Create<AppVersionImpl>(app.version, app.appName, app.category, app.url));
+            }
+            std::list<AppImpl*> apps;
+            for(const auto& app: appsDet)
+            {
+                auto appVersions = app.second;
+                apps.push_back(Core::Service<AppImpl>::Create<AppImpl>(app.first.first, app.first.second, appVersions));
+                for(auto appVersion : appVersions) {
+                    appVersion->Release();
+                }
+            }
+
+            // Create apps payload which will be returned as the result
+            ILISA::IAppsPayload* appsPayload = Core::Service<AppsPayloadImpl>::Create<ILISA::IAppsPayload>(apps);
+            result = appsPayload;
+
+            for (auto app : apps) {
+                app->Release();
```

Comment:
Are you sure we can release these apps here?
The caller of this function will refer to these apps (in returned appsPayload) and that will happen after this function has ended.

Reply:
Yes, every AppImpl from the list is addRef'ed in AppsPayloadImpl constructor.